File: 0064-minimum-path-sum/0064-minimum-path-sum.py
import sys
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        n=len(grid)
        m=len(grid[0])
        dp=[[sys.maxsize for j in range(m+1)]for i in range(n+1)]
        for i in range(n-1,-1,-1):
            for j in range(m-1,-1,-1):
                if i==n-1 and j==m-1:
                    dp[i][j]=grid[i][j]
                    # print(dp[i][j])
                else:
                   


                        ans1=dp[i+1][j]
                        ans2=dp[i][j+1]
                     
                        dp[i][j]=(grid[i][j]+min(ans1,ans2))
                    # print(dp[i][j])
        return dp[0][0]

File: 0064-minimum-path-sum/0064-minimum-path-sum.cpp
class Solution {

private:

    int helper(int i,int j,vector<vector<int>>& grid,vector<vector<int>>&dp){
        if(i==0 and j==0 )return grid[i][j];
        if(i<0 or j<0) return 1e9;
        if(dp[i][j]!=-1) return dp[i][j];
        int up=grid[i][j]+helper(i-1,j,grid,dp);
        int left=grid[i][j]+helper(i,j-1,grid,dp);
        return dp[i][j]=min(up,left);
    }
public:
    int minPathSum(vector<vector<int>>& grid) {

        int n=grid.size();
        int m=grid[0].size();
        vector<vector<int>>dp(n,vector<int>(m,-1));

        return helper(n-1,m-1,grid,dp);
        
    }
};